group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/ConcurrentEventQueue.java

```diff
@@ -0,0 +1,258 @@
+ *
+ * This class is threadsafe.
+ *
+ * @param <K> The type of the key of the event.
+ * @param <T> The type of the event itself. It implements the {{@link Event}} interface.
+ *
+ * There are a few examples about how to use it in the unit tests.
+ */
+public class ConcurrentEventQueue<K, T extends ConcurrentEventQueue.Event<K>> implements AutoCloseable {
+
+    /**
+     * The interface which must be implemented by all events.
+     *
+     * @param <K> The type of the key of the event.
+     */
+    public interface Event<K> {
+        K key();
+    }
+
+    /**
+     * The random generator used by this class.
+     */
+    private Random random;
+
+    /**
+     * The map of queues keyed by K.
+     */
+    private Map<K, Queue<T>> queues;
+
+    /**
+     * The list of available keys. Keys in this list can
+     * be delivered to pollers.
+     */
+    private List<K> availableKeys;
+
+    /**
+     * The set of keys that are being processed.
+     */
+    private Set<K> inflightKeys;
+
+    /**
+     * The number of events in the queue.
+     */
+    private int size;
+
+    /**
+     * A boolean indicated whether the queue is closed.
+     */
+    private boolean closed;
+
+    /**
+     * The lock for protecting access to the resources.
+     */
+    private ReentrantLock lock;
+
+    /**
+     * The condition variable for waking up poller threads.
+     */
+    private Condition condition;
+
+    public ConcurrentEventQueue() {
+        this(new Random());
+    }
+
+    public ConcurrentEventQueue(
+        Random random
+    ) {
+        this.random = random;
+        this.queues = new HashMap<>();
+        this.availableKeys = new ArrayList<>();
+        this.inflightKeys = new HashSet<>();
+        this.closed = false;
+        this.lock = new ReentrantLock();
+        this.condition = lock.newCondition();
+    }
+
+    /**
+     * Adds an {{@link Event}} to the queue.
+     *
+     * @param event An {{@link Event}}.
+     */
+    public void add(T event) {
+        lock.lock();
+        try {
+            if (closed) throw new IllegalStateException("Can't accept an event because the queue is close.");
+
+            K key = event.key();
+            Queue<T> queue = queues.get(key);
+            if (queue == null) {
+                queue = new LinkedList<>();
+                queues.put(key, queue);
+                if (!inflightKeys.contains(key)) {
+                    addAvailableKey(key);
+                }
+            }
+            queue.add(event);
+            size++;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Returns the next {{@link Event}} available. This method block indefinitely until
+     * one event is ready or the queue is closed.
+     *
+     * @return The next event.
+     */
+    public T poll() {
+        return poll(Long.MAX_VALUE, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Returns the next {{@link Event}} available. This method blocks for the provided
```

Comment:
Yeah, I agree with you. I don't like this name either. How about `EventDispatcher`?

Reply:
I was thinking something along `KeyedEventSet` with a `pollAny` method to highlight an element of any given key not already in-use can be taken. One of the reason it is difficult to name is that the class combines a data structure with a semaphore semantic over one attribute (the key) of the elements in that data structure. If round-robin selection could be used (I don't know) instead of random selection, an actual FIFO or LIFO queue of key -> [events] augmented with the set of keys could also be used. But I cannot think about another approach which is as efficient as the bespoke implementation here.

On a separate note, I wonder if the random selection of keys could be extracted and provided to the component to make testing deterministic?

On the `EventDispatcher` name: isn't dispatcher typically used for components which actually performs a dispatch/scheduling action? (e.g. a dispatcher thread)

Reply:
We can also just pass in a seed for random to make testing deterministic right?

Reply:
`EventAccumulator` sounds good. @Justine, yes, you are right, we could also provide a seed.